File: ARSAKA_PUGUH/backend/core/decision/domain/decision.py
"""
Decision Domain Entity

Represents a decision record in the system.
Based on existing decisions table from 001_initial_schema.sql
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional
from uuid import UUID


class DecisionOutcome(str, Enum):
    """Decision outcome enum."""
    ALLOWED = "ALLOWED"
    DENIED = "DENIED"
    REQUIRE_APPROVAL = "REQUIRE_APPROVAL"


@dataclass
class Decision:
    """Decision domain entity (immutable record)."""

    id: UUID
    tenant_id: UUID
    decision_type: str
    context: Dict[str, Any]
    outcome: DecisionOutcome
    rule_matched_id: Optional[UUID] = None
    rule_version: Optional[str] = None
    approval_workflow_id: Optional[UUID] = None
    idempotency_key: Optional[str] = None
    latency_ms: Optional[int] = None
    created_at: Optional[datetime] = None
    metadata: Optional[Dict[str, Any]] = None

    def requires_approval(self) -> bool:
        """Check if decision requires approval workflow."""
        return self.outcome == DecisionOutcome.REQUIRE_APPROVAL

    def is_allowed(self) -> bool:
        """Check if decision was allowed."""
        return self.outcome == DecisionOutcome.ALLOWED

    def is_denied(self) -> bool:
        """Check if decision was denied."""
        return self.outcome == DecisionOutcome.DENIED

File: ARSAKA_PUGUH/frontend/src/features/decisions/types/Decision.ts
/**
 * Decision Types - Phase A+
 * Minimal types for decision visibility
 */

export interface Decision {
  decision_id: string
  tenant_id: string
  decision_type: string
  outcome: 'ALLOWED' | 'DENIED' | 'REQUIRE_APPROVAL'
  outcome_label: string
  rule_matched_id: string | null
  approval_workflow_id: string | null
  created_at: string
  context: Record<string, any>
  latency_ms: number | null
}

export interface DecisionListResponse {
  decisions: Decision[]
  total: number
}

export interface DecisionDetail {
  decision: {
    decision_id: string
    tenant_id: string
    decision_type: string
    context: Record<string, any>
    outcome: string
    outcome_label: string
    outcome_description: string
    rule_matched_id: string | null
    rule_version: string | null
    approval_workflow_id: string | null
    latency_ms: number | null
    created_at: string
    metadata_json: Record<string, any> | null
  }
  workflow: {
    workflow_id: string
    current_state: string
    current_state_label: string
    approver_role: string
    delegated_to_user_id: string | null
    escalated_to_user_id: string | null
    created_at: string
    completed_at: string | null
  } | null
  workflow_actions: Array<{
    action_id: string
    action_type: string
    action_type_label: string
    old_state: string | null
    new_state: string
    acted_by_role: string
    acted_by_user_id: string | null
    action_at: string
    comment: string | null
    metadata_json: Record<string, any> | null
  }>
}

export interface CreateDecisionRequest {
  tenant_id: string
  decision_type: string
  context: Record<string, any>
  idempotency_key?: string
  trace_id?: string
  requester_user_id?: string
}

export interface CreateDecisionResponse {
  decision_id: string
  outcome: string
  rule_matched_id: string | null
  rule_version: string | null
  workflow_id: string | null
  created_at: string
}
